### ISSUE TITLE
[FEATURE REQUEST]: Introduce Partial Indexes to index subset of data based on a filter condition

### ISSUE THREAD

**Issue by samdvr** [microsoft/hyperspace#305]
Introduce Partial Indexes to index subset of data based on a filter condition

As a  user I want to create index on a filtered dataframe with a condition
This filter can be used on read query as a more selective index for access patterns.

ex:
```scala
hs.createIndex(df,IndexConfig("index1", indexedColumns = Seq("id"), condition: "id > 10")
```

**Comment by clee704** [comment]
@rapoth I think it would be an interesting feature to implement. WDYT?

**Comment by rapoth** [comment]
@clee704 I agree this is an interesting feature. I was thinking we could do this as part of #186 since the latter is much more general. What do you think?

**Comment by clee704** [comment]
I think a partial index or filtered index is a different thing, because materialization is not required for them. But as we only have covering indexes for now, there is little difference indeed.
